cpp/test/IceSSL/configuration/OpenSSLTests.cpp

```diff
@@ -0,0 +1,531 @@
+
+    const string serverCertFile = testDir + "/../certs/s_rsa_ca1_pub.pem";
+    const string serverKeyFile = testDir + "/../certs/s_rsa_ca1_priv.pem";
+    SSL_CTX_use_certificate_chain_file(serverSSLContext, serverCertFile.c_str());
+    SSL_CTX_use_PrivateKey_file(serverSSLContext, serverKeyFile.c_str(), SSL_FILETYPE_PEM);
+    SSL_CTX_set_default_passwd_cb(serverSSLContext, passwordCallback);
+
+    // const string clientCAFile = testDir + "/../certs/cacert2.pem";
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_client_method());
+
+    // SSL_CTX_load_verify_file(clientSSLContext, clientCAFile.c_str());
+
+    try
+    {
+        Ice::SSL::ServerAuthenticationOptions serverAuthenticationOptions{
+            .serverSSLContextSelectionCallback = [serverSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(serverSSLContext);
+                return serverSSLContext;
+            }};
+        Ice::CommunicatorHolder serverCommunicator(createServer(serverAuthenticationOptions, helper));
+
+        Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+            .clientSSLContextSelectionCallback = [clientSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(clientSSLContext);
+                return clientSSLContext;
+            }};
+        Ice::CommunicatorHolder clientCommunicator(createClient(clientAuthenticationOptions));
+
+        ServerPrx obj(clientCommunicator.communicator(), "server:" + helper->getTestEndpoint(20, "ssl"));
+        obj->ice_ping();
+        test(false);
+    }
+    catch (const Ice::SecurityException&)
+    {
+        // Expected
+    }
+    catch (...)
+    {
+        SSL_CTX_free(serverSSLContext);
+        SSL_CTX_free(clientSSLContext);
+        throw;
+    }
+    SSL_CTX_free(serverSSLContext);
+    SSL_CTX_free(clientSSLContext);
+    cout << "ok" << endl;
+}
+
+void
+clientRejectsServerUsingDefaultSettings(Test::TestHelper* helper, const string& testDir)
+{
+    cout << "client rejects server using default settings... " << flush;
+    SSL_CTX* serverSSLContext = SSL_CTX_new(TLS_server_method());
+
+    const string serverCertFile = testDir + "/../certs/s_rsa_ca1_pub.pem";
+    const string serverKeyFile = testDir + "/../certs/s_rsa_ca1_priv.pem";
+    SSL_CTX_use_certificate_chain_file(serverSSLContext, serverCertFile.c_str());
+    SSL_CTX_use_PrivateKey_file(serverSSLContext, serverKeyFile.c_str(), SSL_FILETYPE_PEM);
+    SSL_CTX_set_default_passwd_cb(serverSSLContext, passwordCallback);
+
+    try
+    {
+        auto serverAuthenticationOptions = Ice::SSL::ServerAuthenticationOptions{
+            .serverSSLContextSelectionCallback = [serverSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(serverSSLContext);
+                return serverSSLContext;
+            }};
+        Ice::CommunicatorHolder serverCommunicator(createServer(serverAuthenticationOptions, helper));
```

Comment:
For the client tests I think only the call that we're expecting to fail should be inside the try catch. Since the only thing we're checking is `Ice::SecurityException`, it could be triggered by any number of issues.